subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit5/JUnitPlatformTestClassProcessor.java

```diff
@@ -0,0 +1,236 @@
+package org.gradle.api.internal.tasks.testing.junit5;
+
+import org.gradle.api.internal.tasks.testing.*;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+import org.gradle.api.tasks.testing.TestResult;
+import org.gradle.internal.time.Clock;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.TestSource;
+import org.junit.platform.engine.reporting.ReportEntry;
+import org.junit.platform.engine.support.descriptor.ClassSource;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+import org.junit.platform.launcher.*;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.launcher.core.LauncherFactory;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static org.junit.platform.engine.discovery.DiscoverySelectors.selectClass;
+
+public class JUnitPlatformTestClassProcessor implements TestClassProcessor {
+    private final Clock clock;
+    private Launcher launcher;
+
+    public JUnitPlatformTestClassProcessor(Clock clock) {
+        this.clock = clock;
+    }
+
+    @Override
+    public void startProcessing(TestResultProcessor resultProcessor) {
+        this.launcher = LauncherFactory.create();
+        launcher.registerTestExecutionListeners(new GradleTestExecutionListener(new InspectingTestResultProcessor(resultProcessor), clock));
+    }
+
+    @Override
+    public void processTestClass(TestClassRunInfo testClass) {
+        System.out.println("Start processing test class: " + testClass.getTestClassName());
+        LauncherDiscoveryRequest request = LauncherDiscoveryRequestBuilder.request()
+            .selectors(selectClass(testClass.getTestClassName()))
+            .build();
+        launcher.execute(request);
```

Comment:
Test classes should be collected and executed in a single run.

Reply:
Doing it as one run is my preference (see my stab at that in #3887). I find the existing testing infrastructure excessively overcomplicated. The JUnit Platform API is wonderful and has a very nice abstraction around it's feature sets. I'd really prefer not to box JUnit Platform support in just to fit the existing Gradle infrastructure.

Reply:
Thanks for your comment on this, @marcphilipp. I was about to that that comment as well.

Reply:
> I'd really prefer not to box JUnit Platform support in just to fit the existing Gradle infrastructure.

I truly could not agree more with that statement!